src/OSPSuite.Core/Domain/Services/KeywordReplacerTask.cs

```diff
@@ -175,34 +173,47 @@ private void replaceInEventGroup(EventGroup eventGroup, KeywordReplacerCollectio
          eventGroup.GetAllChildren<EventAssignment>().Select(x => x.ObjectPath).Each(keywordReplacer.ReplaceIn);
       }
 
-      public ObjectPath CreateModelPathFor(ObjectPath objectPath, IContainer rootContainer)
+      public ObjectPath CreateModelPathFor(ObjectPath objectPath, ReplacementContext replacementContext)
       {
          var keywordReplacer = new KeywordReplacerCollection();
-         addCommonModelReplacersTo(keywordReplacer, rootContainer);
+         addCommonModelReplacersTo(keywordReplacer, replacementContext);
          var modelPath = objectPath.Clone<ObjectPath>();
          keywordReplacer.ReplaceIn(modelPath);
          return modelPath;
       }
 
-      public void ReplaceIn(IContainer rootContainer) => replaceInContainer(rootContainer, rootContainer);
+      public void ReplaceIn(ModelConfiguration modelConfiguration)
+      {
+         //When we validate the whole configuration, we update the replacement context to make sure that the replacement context is up to date
+         modelConfiguration.UpdateReplacementContext();
+         replaceInContainer(modelConfiguration.Model.Root, modelConfiguration.ReplacementContext);
+      }
 
-      private void replaceInContainer(IContainer container, IContainer rootContainer)
+      private void replaceInContainer(IContainer container, ReplacementContext replacementContext)
       {
-         try
-         {
-            _rootContainer = rootContainer;
-            container.AcceptVisitor(this);
-         }
-         finally
+         //neighborhood replacements
+         container.GetAllChildren<Neighborhood>().Each(x => ReplaceIn(x, replacementContext));
```

Comment:
For the speedup:
a) this code line could be moved into `public void ReplaceIn(ModelConfiguration modelConfiguration)` 
    All other places where `replaceInContainer()` is called deal with reactions, transport realizations, etc. and will retrieve no nbhoods
b) when moved: we could get nbhoods list explicitly instead of calling `GetAllChildren<Neighborhood>()`

Reply:
yeah interesting. I did not change any of the logic but I think you are right. Let's try :)

Reply:
ok